File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
from collections import deque

def solution(numbers, target):
    cnt = 0
    n = len(numbers)
    queue = deque()
    queue.append([numbers[0], 0])
    queue.append([-1*numbers[0], 0])
    
    while queue:
        tmp, idx = queue.popleft()
        idx += 1
        if idx < n: # 계산할 값이 남음
            queue.append([tmp + numbers[idx], idx])
            queue.append([tmp - numbers[idx], idx])
        else: # 계산할 값이 없음
            if tmp == target:
                cnt += 1
    return cnt

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    int cnt = 0;
    public int solution(int[] numbers, int target) {
        Calc(numbers, target, 0, 0);
        return cnt;
    }
    
    public void Calc(int[] numbers, int target, int sum, int idx) {
        if (idx == numbers.length) {
            if (sum == target) cnt++;
            return;
        }
        
        Calc(numbers, target, sum + numbers[idx], idx + 1);
        Calc(numbers, target, sum - numbers[idx], idx + 1);
    }
}